File: algorithms/I/Implement Magic Dictionary/Implement Magic Dictionary.cpp
struct node{
	bool end = false;
	node *children[26];
};

class MagicDictionary {
public:

	node* root;
	void insert(string&s){
		node* cur = root;
		for(char&c : s){
			if(!cur->children[c-'a']){
				cur->children[c-'a'] = new node();
			}
			cur = cur->children[c-'a'];
		}
		cur->end=true;
	}
	MagicDictionary() {
		root = new node();
	}

	void buildDict(vector<string> dictionary) {
		for(string&s : dictionary)insert(s);
	}
	
	bool find(int i, string& s,node* cur, int mismatch){
		if(i == s.size() and mismatch==0) return cur->end;
		if(mismatch<0)return false;
		if((!cur || i == s.size()) and mismatch!=0) return false;

		int ind = s[i]-'a';
		bool ans = false;
		for(int j=0;j<26;j++){
			if(cur->children[j]){
				if(j!=ind){
					ans |= find(i+1,s,cur->children[j],mismatch-1);
				}else{
					ans |= find(i+1,s,cur->children[j],mismatch);
				}
			}
		}
		return ans;
	}
	bool search(string searchWord) {
		return find(0,searchWord,root,1);
	}
};

File: algorithms/I/Implement Magic Dictionary/Implement Magic Dictionary.py
class MagicDictionary:

	def __init__(self):
		TrieNode = lambda : defaultdict(TrieNode)
		self.root = TrieNode()

	def buildDict(self, dictionary: List[str]) -> None:
		for s in dictionary:
			cur = self.root
			for c in s: cur = cur[ord(c)-ord('a')]
			cur['$']=True

	def search(self, searchWord: str) -> bool:
		def find(i,cur,mis):
			if i==len(searchWord) and mis==0: return('$' in cur)
			if mis < 0: return False
			if i==len(searchWord) and mis!=0: return False
			ind = ord(searchWord[i])-ord('a')
			ans = False
			for j in range(26):
				if j in cur:
					if(j!=ind):
						ans |= find(i+1,cur[j],mis-1)
					else: ans |= find(i+1,cur[j],mis)
			return ans
			
		return find(0,self.root,1)
